Add Media Group object

dromo77 opened the pull request:
## Overview

This PR adds a Media Group object. It should be helpful for aligning images/content on pages like the [Case Study Listing](https://v-next--cloudfour-patterns.netlify.app/iframe.html?id=prototypes-case-study-listings--example&args=&viewMode=story) page.

## Screenshots

<img width="1166" alt="Screen Shot 2021-11-11 at 11 12 16 AM" src="https://user-images.githubusercontent.com/42841342/141355416-0bd28613-713d-4be8-ad3d-e4784c0359a6.png">

## Testing

1.

---

- Fixes #1390

tylersticka commented:
@dromo77 Thanks for tackling this challenging layout problem.

I'm not sure I understand what `o-media-group__item` does that `o-media` does not, or what `o-media-group` does that `o-deck` does not. I assume I'm missing something between the way the two interact?

dromo77 commented:
> I'm not sure I understand what `o-media-group__item` does that `o-media` does not, or what `o-media-group` does that `o-deck` does not. I assume I'm missing something between the way the two interact?

@tylersticka I ran into a couple issues trying to use `o-media` and `o-deck` as a solution. In the prototype the content splits into two columns at the `l` breakpoint.  `o-deck` automatically creates columns, so there were multiple columns before the `l` breakpoint. The content looked a little squished.

`o-media` doesn't size the object and also uses `flex`. I thought about some sort of modifier that uses CSS Grid. This would help size of the object and could also be used to align objects vertically, similar to what the `o-media-group__item` is doing. However, the alignment and grid size of the object are different on small screens versus large screens. I wasn't sure how to handle the different sizes/alignment needed. 

I'm open to other solutions, but wanted to give some context to the problems I ran into using existing patterns.

tylersticka commented:
> `o-deck` automatically creates columns, so there were multiple columns before the `l` breakpoint. The content looked a little squished.

I assume that was true even with the `o-deck--2-column@l`modifier. Perhaps we should add `1-column` modifiers to `o-deck`? Would that solve the issue?

> `o-media` doesn't size the object and also uses `flex`. I thought about some sort of modifier that uses CSS Grid. This would help size the object and could also be used to align objects vertically, similar to what the `o-media-group__item` is doing. However, the alignment and grid size of the object are different on small screens versus large screens. I wasn't sure how to handle the different sizes/alignment needed.

Reading the comment in the `media.scss` file, it seems like I used Flexbox to write it just because it was simple. But that's not a very compelling reason to _keep_ using Flexbox. If there's a good reason to use CSS Grid instead, we should totally do that.

From there, there are probably a few ways of handling those case study image widths. Perhaps we could make one or more `o-media` modifiers that enforce a certain object column width?

Or, honestly, I don't mind the different media columns being different as much as I used to. If we find that's taking an outsized amount of effort to componentize, maybe it isn't completely necessary after all? What do you think?

dromo77 commented:
>  If we find that's taking an outsized amount of effort to componentize, maybe it isn't completely necessary after all?

I'll try using our existing patterns with the changes you mentioned. I don't think it will take too much time to add `o-media` modifiers, but if I find otherwise we can consider different sized columns.

dromo77 commented:
@tylersticka I made several changes:

1. I updated the [Case Study Listing](https://deploy-preview-1579--cloudfour-patterns.netlify.app/?path=/story/prototypes-case-study-listings--example) prototype to use patterns so we can see what it looks like in action.
2. I updated the `o-deck` column modifier so you can specify when you want 1 column.
3. I updated `o-media` to use CSS grid when supported. I _think_ I did it correctly. I haven't noticed any regressions, but would appreciate another pair of eyes on it.
4.  I added a modifier to specify the size of `o-media__object` relative to `o-media__content`. I want to make sure the modifier name and documentation makes sense.
5.  I added another modifier to `o-media` to fix some alignment issues I noticed in the prototype.

tylersticka commented on .changeset/polite-gorillas-wave.md:
```diff
@@ -0,0 +1,5 @@
+'@cloudfour/patterns': minor
+---
+
+Update Deck columns modifier. Update Media to use CSS Grid and add modifier to size content.
```
1. Can we add these as two separate changelog updates?
2. Can we be more specific about what modifiers were added?
3. Can we add a bit of "why" information?

tylersticka commented on src/objects/media/media.scss:
```diff
@@ -50,3 +52,49 @@
+  }
+
+  @for $i from 1 through 4 {
+    .o-media--1by#{$i} {
```
I like this class name, but can we add hyphens so it's `--1-by-X` for consistency with our other class names?

tylersticka commented on src/objects/media/media.scss:
```diff
@@ -50,3 +52,49 @@
+    }
+  }
+
+  /**
+   * Reset fallback styles, we use `grid-gap` when CSS Grid is supported
+   */
+  .o-media__object {
+    margin-right: 0;
+  }
+
+  .o-media--reverse .o-media__object {
+    margin-left: 0;
+  }
```
I don't think these are necessary since Grid is so well supported now.

tylersticka commented on src/objects/media/media.scss:
```diff
@@ -50,3 +52,49 @@
+ * We use CSS grid when supported by the browser.
+ */
+
+@supports (display: grid) {
```
Since IE11 is the only browser that doesn't support Grid these days, I wonder if having a fallback is still necessary?

tylersticka commented on src/objects/media/media.scss:
```diff
@@ -50,3 +52,49 @@
+ * Items remain top aligned
+ */
+
+.o-media--alignStart {
```
For consistency with our BEMIT class names, this should be `--align-start`

tylersticka commented on src/objects/media/media.stories.mdx:
```diff
@@ -93,6 +94,44 @@ The order of `o-media__object` and `o-media__content` will not affect this patte
 
+## Alignment
+
+By default, content will appear middle-aligned with the object. If the content exceeds the object's height, they will appear top-aligned. You can change the alignment so the content and object remain top aligned using the modifier `o-media--alignStart`.
```
It isn't very clear to me _why_ we would do this? 🤔

dromo77 replied:
When I updated the Case Study Listing prototype I noticed some of the media objects and content were top-aligned and some were middle-aligned. In a group like this, it looks off when the alignment is different.

![Screen Shot 2021-11-12 at 1 18 14 PM](https://user-images.githubusercontent.com/42841342/141537289-fd64fa0c-2178-4cd3-8574-4abb303e67bb.png)

tylersticka replied:
Ah, interesting. I wonder if that's better solved on the media object or on the deck by changing `align-items` or `align-content` via a modifier?

dromo77 replied:
Hmm, that's a good question. I don't have a strong opinion on whether we add a modifier to media or the deck. What do you think?

tylersticka replied:
I think it makes more sense on `o-deck` for a couple of reasons:

- The `o-media` object isn't really causing this alignment issue on its own: It's because the Deck is causing it to stretch to fill the column.
- I think `o-deck--align-start` is easier to understand than `o-media--align-start`, because to most people I think they'd look at `o-media` and think "isn't that already aligned to the start?" Whereas if you were to have a deck of cards I think it would be easier to see the difference.
- It's a little nicer to put the modifier on one container instead of every child element.

tylersticka reviewed (commented):
It looks like some of the prototype images have been copied from `src/prototypes/case-study-listings/images` to `static/media`. That's fine, but can we remove the duplicates so they are only in one place?

tylersticka reviewed (commented):
Apparently I wrote a comment on Friday that I never actually posted. Sorry about that. 🤦

dromo77 commented on src/objects/deck/deck.stories.mdx:
```diff
@@ -83,11 +84,25 @@ By default, the `o-deck` class will use CSS Grid Layout to arrange child element
+  >
+    {articlesStory.bind({})}
+  </Story>
+</Canvas>
```
@tylersticka I updated the example, let me know if it makes more sense 🙂
